base/applications/games/solitaire/solundo.cpp

```diff
@@ -0,0 +1,149 @@
+#include "solitaire.h"
+
+static int source_id = -1;
+static int destination_id = -1;
+static int number_of_cards = 0;
+static int prev_score = 0;
+static int prev_visible_pile_cards = 0;
+
+void SetUndo(
+    int set_source_id,
+    int set_destination_id,
+    int set_number_of_cards,
+    int set_prev_score,
+    int set_prev_visible_pile_cards)
+{
+    if (set_source_id == set_destination_id)
+        return;
+
+    source_id = set_source_id;
+    destination_id = set_destination_id;
+    number_of_cards = set_number_of_cards;
+    prev_score = set_prev_score;
+    prev_visible_pile_cards = set_prev_visible_pile_cards;
+    SetUndoMenuState(true);
+}
+
+void ClearUndo(void)
+{
+    source_id = -1;
+    destination_id = -1;
+    number_of_cards = 0;
+    SetUndoMenuState(false);
+}
+
+void Undo(void)
+{
+    CardRegion *source = NULL;
+    CardRegion *destination = NULL;
+
+    if (source_id < 1 || destination_id < 1 || number_of_cards < 1)
+    {
+        ClearUndo();
+        return;
+    }
+
+    if (source_id >= ROW_ID)
+        source = pRowStack[source_id - ROW_ID];
+    else if (source_id >= SUIT_ID)
+        source = pSuitStack[source_id - SUIT_ID];
+    else if (source_id == PILE_ID)
+        source = pPile;
+    else if (source_id == DECK_ID)
+        source = pDeck;
+
+    if (destination_id >= ROW_ID)
```

Comment:
Same comment for destination_id as above.

Reply:
It's also done.